src/fifo.c

```diff
@@ -0,0 +1,353 @@
+         *  an index into the block indicating the first or last item in the block, depending on
+         *  context.
+         *
+         * This UNION overlays a pointer with an integral value.  This allows us to look at the
+         *  pointer OR the integer without casting - but they use the same memory.
+         *
+         * If there is MORE THAN ONE block in the chain, the first block has a pointer/index that
+         *  looks like this.  However, if there is only a single block, it looks like the LAST block.
+         *   +-----------------------------------------------------------+
+         *   |                 next pointer                   | firstIdx |
+         *   |                  (61 bits)                     | (3 bits) |
+         *   +-----------------------------------------------------------+
+         *     * The next pointer is only valid after zeroing out the last 3 bits.
+         *     * "lastIdx" is implied to be 6 (because there are additional blocks).
+         *     * "firstIdx" represents the first filled index (0..6).  POP occurs here.
+         *
+         * Any blocks in the middle of the chain have a regular pointer like this:
+         *   +-----------------------------------------------------------+
+         *   |                 next pointer                   |    0*    |
+         *   |                  (61 bits)                     | (3 bits) |
+         *   +-----------------------------------------------------------+
+         *     * The next pointer is valid as-is
+         *     * "lastIdx" is implied to be 6 in all middle blocks.
+         *     * "firstIdx" is implied to be 0 in all middle blocks.
+         *     * NOTE: In middle blocks, the index bits(0) are really still the firstIdx value.
+         *             When Fifo's are joined, the O(1) operation may result in a partially
+         *             full middle block.  In this case, the items are "right-justified" and
+         *             firstIdx indicates where the items start.
+         *
+         * The last (or only) block in the chain contains only the lastIndex, the pointer is unused.
+         *   +-----------------------------------------------------------+
+         *   |                      0                         | lastIdx  |
+         *   |                  (61 bits)                     | (3 bits) |
+         *   +-----------------------------------------------------------+
+         *     * The next pointer is unused and guaranteed NULL.
+         *     * "lastIdx" represents the last filled index (0..6).
+         *     * "firstIdx" is implied to be zero on the last (or only) block.
+         */
+        uintptr_t last_or_first_idx;
+        fifoBlock *next;
+    } u;
+};
+
+struct fifo {
+    long length; /* Number of items */
+    fifoBlock *first;
+    fifoBlock *last;
+};
+
+
+fifo *fifoCreate(void) {
+    fifo *q = zmalloc(sizeof(fifo));
+    q->length = 0;
+    q->first = q->last = NULL;
+    return q;
```

Comment:
```suggestion
    fifo *f = zmalloc(sizeof(fifo));
    f->length = 0;
    f->first = f->last = NULL;
    return f;
```
Not a major point, but using q makes even less sense now.

Reply:
Since fifo IS a queue, `q` makes sense and is readable. We’d rather not change that.